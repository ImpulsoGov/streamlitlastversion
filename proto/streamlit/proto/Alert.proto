/**
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

syntax = "proto3";

message Alert {
  // Content to display.
  string body = 1;

  // Type of Alert
  enum Format {
    // Plain, fixed width text.
    UNUSED = 0;

    // Shows an error message.
    ERROR = 1;

    // Shows a warning message.
    WARNING = 2;

    // Shows an info log.
    INFO = 3;

    // Shows a success message.
    SUCCESS = 4;
  }
  Format format = 2;
}
