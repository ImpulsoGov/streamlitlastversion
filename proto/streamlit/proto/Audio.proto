/**
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

syntax = "proto3";

message Audio {
  string url = 5;

  // The currentTime attribute of the HTML <audio> tag's <source> subtag.
  int32 start_time = 3;

  reserved 1, 2, 4;
  reserved "data", "format";
}
