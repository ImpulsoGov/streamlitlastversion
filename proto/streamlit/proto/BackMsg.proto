/**
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

syntax = "proto3";

import "streamlit/proto/Widget.proto";

// A message from the browser to the server.
message BackMsg {
  oneof type {
    // Set to true to ask the server to display the inline help.
    // bool help = 1;  DEPRECATED

    // Set to true to ask the server to upload the report.
    bool cloud_upload = 2;

    // Asks the server to run the script.
    bool rerun_script = 3;

    // DEPRECATED - Asks the server to run the script with this object
    // ReRun rerun = 4;

    // Requests that the report's @st_cache be cleared
    bool clear_cache = 5;

    // Requests that the runOnSave behavior for this report be set
    // to the given value
    bool set_run_on_save = 6;

    // Requests that the report's execution be stopped
    bool stop_report = 7;

    // string widget_json = 8;

    WidgetStates update_widgets = 9;

    // Set to true to ask the server to close the connection
    bool close_connection = 10;
  }
}
