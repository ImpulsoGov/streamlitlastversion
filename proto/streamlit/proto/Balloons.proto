/**
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

syntax = "proto3";

// A python empty.
message Balloons {
  Type type = 1;

  // A random number. If same as previous run, will not reshow balloons.
  uint32 execution_id = 2;

  enum Type {
    DEFAULT = 0;
    BALLOON = 1;
    HAPPY_FACE = 2;
    STAR_FACE = 3;
    COOL_FACE = 4;
  }
}
