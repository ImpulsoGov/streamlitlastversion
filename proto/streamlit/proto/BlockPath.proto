/**
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

syntax = "proto3";

message BlockPath {
  enum Container {
    UNUSED = 0;
    MAIN = 1;
    SIDEBAR = 2;
  }

  // Top level container of the block.
  Container container = 1;

  // A list with the indices of all ancestors of the block.
  repeated uint32 path = 2;
}
