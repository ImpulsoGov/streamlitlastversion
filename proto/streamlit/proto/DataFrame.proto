/**
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

syntax = "proto3";

// Represents a pandas DataFrame.
message DataFrame {
  // The data in the array.
  Table data = 1;

  // List of row names. (Multiple implies a multi-index.)
  Index index = 2;

  // List of column names. (Multiple implies a multi-index.)
  Index columns = 3;

  // Cell style and formatting data. Optional.
  TableStyle style = 4;
}

// An index in the dataFrame
message Index {
  oneof type {
    PlainIndex plain_index = 1;
    RangeIndex range_index = 2;
    // CategoricalIndex categorical_index = 3;
    MultiIndex multi_index = 4;
    // IntervalIndex interval_index = 5;
    DatetimeIndex datetime_index = 6;
    TimedeltaIndex timedelta_index = 7;
    // PeriodIndex period_index = 8;
    Int64Index int_64_index = 9;
    // UInt64Index uint_64_index = 10;
    Float64Index float_64_index = 11;
  }
}

// Basic, 1D index.
message PlainIndex {
  AnyArray data = 1;

  //// Not yet implemented:
  // DType dtype = 2;
  // bool copy = 3;
  // string name = 4;
  // bool tupleize_cols = 5;
}

// Range index. See:
// https://pandas.pydata.org/pandas-docs/stable/generated/pandas.RangeIndex.html
message RangeIndex {
  int64 start = 1;
  int64 stop = 2;

  //// Not yet implemented:
  // string name = 4;
  // bool copy = 5;
}

// A multi-level, or hierarchical, Index. See:
// https://pandas.pydata.org/pandas-docs/stable/generated/pandas.MultiIndex.html
message MultiIndex {
  repeated Index levels = 1;
  repeated Int32Array labels = 2;

  //// Not yet implemented:
  // int sort_order = 3;
  // StringArray names = 4;
  // bool copy = 5;
  // bool verify_integrity = 6;
}

// A date represented internally as nano second epoch int64. See:
// https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DatetimeIndex.html
message DatetimeIndex {
  Int64Array data = 1;
}

// A time interval represented internally as nano second epoch int64. See:
// https://pandas.pydata.org/pandas-docs/stable/generated/pandas.TimedeltaIndex.html
message TimedeltaIndex {
  Int64Array data = 1;
}

// See:
// https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Int64Index.html
message Int64Index {
  Int64Array data = 1;
}

// See:
// https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Int64Index.html
message Float64Index {
  DoubleArray data = 1;
}

message StringArray {
  repeated string data = 1;
}

message DoubleArray {
  repeated double data = 1;
}

message Int32Array {
  repeated int32 data = 1;
}

message Int64Array {
  repeated int64 data = 1;
}

message UInt32Array {
  repeated uint32 data = 1;
}

message CSSStyle {
  string property = 1;
  string value = 2;
}

message CellStyle {
  repeated CSSStyle css = 1;
  string display_value = 2;       // e.g. '21.96%'

  // The default value for a string field in proto3 is '', so we need
  // this extra bool to indicate the presence of a user-specified
  // display_value, which itself could be ''.
  bool has_display_value = 3;
}

message CellStyleArray {
  repeated CellStyle styles = 1;
}

message AnyArray {
  oneof type {
    StringArray strings = 1;
    DoubleArray doubles = 2;
    Int64Array int64s = 3;
    Int64Array datetimes = 4;
    Int64Array timedeltas = 5;
  }
}

message Table {
  repeated AnyArray cols = 1;
}

message TableStyle {
  repeated CellStyleArray cols = 1;
}
