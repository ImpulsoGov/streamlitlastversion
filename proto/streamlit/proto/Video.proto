/**
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

syntax = "proto3";

message Video {
  // A url pointing to a video file
  string url = 6;

  enum Type {
    UNUSED = 0;  // This should always exist.
    NATIVE = 1;
    YOUTUBE_IFRAME = 2;
  }

  // The currentTime attribute of the HTML <video> tag's <source> subtag.
  int32 start_time = 3;
  
  // Type affects how browser wraps the video in tags: plain HTML5, YouTube...
  Type type = 5;

  reserved 1,2,4;
  reserved "format", "data";
}
