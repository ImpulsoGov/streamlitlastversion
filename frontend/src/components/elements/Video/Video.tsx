/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { createRef, PureComponent, ReactNode } from "react"
import { Map as ImmutableMap } from "immutable"
import { Video as VideoProto } from "autogen/proto"
import {
  buildHttpUri,
  getWindowBaseUriParts as get_base_uri_parts,
} from "lib/UriUtil"

export interface Props {
  width: number
  element: ImmutableMap<string, any>
}

class Video extends PureComponent<Props> {
  private videoRef = createRef<HTMLVideoElement>()

  public componentDidMount = (): void => {
    this.updateTime()
  }

  public componentDidUpdate = (): void => {
    this.updateTime()
  }

  private updateTime = (): void => {
    if (this.videoRef.current) {
      const { element } = this.props

      this.videoRef.current.currentTime = element.get("startTime")
    }
  }

  private getYoutubeSrc = (): string => {
    const { element } = this.props
    const startTime = element.get("startTime")
    const url = element.get("url")

    if (startTime) {
      return `${url}?start=${startTime}`
    }

    return url
  }

  private getVideoURI = (): string => {
    const { element } = this.props
    const url = element.get("url")

    if (url.startsWith("/media")) {
      return buildHttpUri(get_base_uri_parts(), url)
    }

    return url
  }

  public render(): ReactNode {
    /* Element may contain "url" or "data" property. */

    const { element, width } = this.props
    const url = element.get("url")
    const type = element.get("type")

    /* Is this a YouTube link? If so we need a fancier tag.
       NOTE: This part assumes the URL is already an "embed" link.
    */
    if (type === VideoProto.Type.YOUTUBE_IFRAME) {
      const height = width * 0.75

      return (
        <iframe
          title={url}
          src={this.getYoutubeSrc()}
          width={width}
          height={height}
          frameBorder="0"
          allow="autoplay; encrypted-media"
          allowFullScreen
        />
      )
    }

    return (
      <video
        ref={this.videoRef}
        controls
        src={this.getVideoURI()}
        className="stVideo"
        style={{ width }}
      />
    )
  }
}

export default Video
