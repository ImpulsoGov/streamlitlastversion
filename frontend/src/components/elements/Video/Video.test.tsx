/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react"
import { shallow } from "enzyme"
import { fromJS } from "immutable"
import { Video as VideoProto } from "autogen/proto"

import Video, { Props } from "./Video"

const getProps = (elementProps: Partial<VideoProto> = {}): Props => ({
  element: fromJS({
    url: "https://www.w3schools.com/html/mov_bbb.mp4",
    type: VideoProto.Type.UNUSED,
    startTime: 0,
    ...elementProps,
  }),
  width: 0,
})

describe("Video Element", () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it("renders without crashing", () => {
    const props = getProps()
    const wrapper = shallow(<Video {...props} />)

    expect(wrapper.find("video").length).toBe(1)
  })

  it("should have correct style", () => {
    const props = getProps()
    const wrapper = shallow(<Video {...props} />)
    const videoWrapper = wrapper.find("video")

    expect(videoWrapper.prop("className")).toContain("stVideo")
    expect(videoWrapper.prop("style")).toStrictEqual({
      width: props.width,
    })
  })

  it("should have controls", () => {
    const props = getProps()
    const wrapper = shallow(<Video {...props} />)

    expect(wrapper.find("video").prop("controls")).toBeDefined()
  })

  it("should build url correctly if it starts with /media", () => {
    const props = getProps({
      url: "/media/url.test.mp4",
    })
    const wrapper = shallow(<Video {...props} />)

    expect(wrapper.find("video").prop("src")).toBe(
      "http://localhost:80/media/url.test.mp4"
    )
  })

  describe("YouTube", () => {
    it("should render a youtube iframe", () => {
      const props = getProps({
        type: VideoProto.Type.YOUTUBE_IFRAME,
      })
      const wrapper = shallow(<Video {...props} />)
      const iframeWrapper = wrapper.find("iframe")

      expect(iframeWrapper.props()).toMatchSnapshot()
    })

    it("should render a youtube iframe with an starting time", () => {
      const props = getProps({
        type: VideoProto.Type.YOUTUBE_IFRAME,
        startTime: 10,
      })
      const wrapper = shallow(<Video {...props} />)
      const iframeWrapper = wrapper.find("iframe")

      expect(iframeWrapper.props()).toMatchSnapshot()
    })
  })

  describe("updateTime", () => {
    const props = getProps()
    const wrapper = shallow(<Video {...props} />)
    const instance = wrapper.instance()

    // @ts-ignore
    jest.spyOn(instance, "updateTime")

    it("should call it on didMount", () => {
      // @ts-ignore
      instance.componentDidMount()

      // @ts-ignore
      expect(instance.updateTime).toBeCalledTimes(1)
    })

    it("should call it on didUpdate", () => {
      // @ts-ignore
      instance.componentDidUpdate()

      // @ts-ignore
      expect(instance.updateTime).toBeCalledTimes(1)
    })
  })
})
