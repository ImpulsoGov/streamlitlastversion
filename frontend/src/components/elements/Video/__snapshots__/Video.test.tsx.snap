// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video Element YouTube should render a youtube iframe 1`] = `
Object {
  "allow": "autoplay; encrypted-media",
  "allowFullScreen": true,
  "frameBorder": "0",
  "height": 0,
  "src": "https://www.w3schools.com/html/mov_bbb.mp4",
  "title": "https://www.w3schools.com/html/mov_bbb.mp4",
  "width": 0,
}
`;

exports[`Video Element YouTube should render a youtube iframe with an starting time 1`] = `
Object {
  "allow": "autoplay; encrypted-media",
  "allowFullScreen": true,
  "frameBorder": "0",
  "height": 0,
  "src": "https://www.w3schools.com/html/mov_bbb.mp4?start=10",
  "title": "https://www.w3schools.com/html/mov_bbb.mp4",
  "width": 0,
}
`;
